### PR TITLE
Cache: support redis cache backend

### PR DESCRIPTION
**Pull request by hanjm**
Hi, I am attempt to add redis cache support,  like cortex does.
First. i will make it a draft PR.

Signed-off-by: Jimmie Han <hanjinming@outlook.com>

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

1. add redis support for bucket-cache/index-cache/query-frontend
## Verification

1. unit test.
2. continning test in production deployment.

### CODE REVIEW

**Review by GiedriusS — COMMENTED** [review]
I haven't checked the code yet but have you seen the implementation on Cortex side https://github.com/cortexproject/cortex/blob/34bb142b129bd781539001fb1629cf2f3ed54b82/pkg/chunk/cache/redis_cache.go? Maybe there is an opportunity to share some code here?

**Comment by hanjm** [comment]
> I haven't checked the code yet but have you seen the implementation on Cortex side https://github.com/cortexproject/cortex/blob/34bb142b129bd781539001fb1629cf2f3ed54b82/pkg/chunk/cache/redis_cache.go? Maybe there is an opportunity to share some code here?

Thank you.your review,  I see the redis code in cortex [not have a expire ttl in store function](https://github.com/cortexproject/cortex/blob/34bb142b129bd781539001fb1629cf2f3ed54b82/pkg/chunk/cache/redis_cache.go#L98)，it use a fixed ttl.  so it can not reuse for our interface https://github.com/thanos-io/thanos/blob/main/pkg/cache/cache.go#L16. the redis support code is not very complex，i think write it again is seems not matter. (:

**Comment by hanjm** [comment]
I found the redis client invoke mget has too many keys. let me implement split it to batch.

**Comment by hanjm** [comment]
🚀  Ready to review. I am also deploymented it to production to continning verify.

File: pkg/queryfrontend/cache.go [hunk 1]
```diff
@@ -25,20 +25,21 @@ func newThanosCacheKeyGenerator(interval time.Duration) thanosCacheKeyGenerator
...
+// TODO(yeya24): Add other request params as request key.
+func (t thanosCacheKeyGenerator) GenerateCacheKey(userID string, r queryrange.Request) string {
 	currentInterval := r.GetStart() / t.interval.Milliseconds()
 	switch tr := r.(type) {
 	case *ThanosQueryRangeRequest:
 		i := 0
 		for ; i < len(t.resolutions) && t.resolutions[i] > tr.MaxSourceResolution; i++ {
 		}
-		return fmt.Sprintf("%s:%d:%d:%d", tr.Query, tr.Step, currentInterval, i)
+		// Cache key should has a uniq prefix, We use `fe` represent frontend.
```
Comment (bwplotka) [inline pkg/queryfrontend/cache.go:37]:
```suggestion
		// Cache key should have a unique prefix. `fe` represents frontend.
```

**Review by bwplotka — APPROVED** [review]
Great job! Just comment to fix, but not a blocker, we can fix later. LGTM! 💪🏽

**Comment by bwplotka** [comment]
Retried e2e test, let's see if that's a flakiness.

**Comment by hanjm** [comment]
Thank you🚀, I will fix it later.

File: pkg/cacheutil/redis_client.go [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+			config.MaxGetMultiConcurrency,
+		),
+		setMultiGate: gate.New(
+			extprom.WrapRegistererWithPrefix("thanos_redis_setmulti_", reg),
+			config.MaxSetMultiConcurrency,
+		),
+	}
+	duration := promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "thanos_redis_operation_duration_seconds",
+		Help:    "Duration of operations against memcached.",
```
Comment (colega) [inline pkg/cacheutil/redis_client.go:171]: Typo: should be _Duration of operations against **redis**._

File: pkg/store/cache/caching_bucket_factory.go [hunk 3]
```diff
@@ -97,6 +98,12 @@ func NewCachingBucketFromYaml(yamlContent []byte, bucket objstore.Bucket, logger
 		if err != nil {
 			return nil, errors.Wrapf(err, "failed to create inmemory cache")
 		}
+	case string(RedisBucketCacheProvider):
+		redisCache, err := cacheutil.NewRedisClient(logger, "caching-bucket", backendConfig, reg)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create memcached client")
```
Comment (colega) [inline pkg/store/cache/caching_bucket_factory.go:104]: Typo: this should say **redis** too.

File: pkg/store/cache/memcached.go [hunk 4]
```diff
@@ -53,15 +53,15 @@ func NewMemcachedIndexCache(logger log.Logger, memcached cacheutil.MemcachedClie
 	c.hits.WithLabelValues(cacheTypePostings)
 	c.hits.WithLabelValues(cacheTypeSeries)
 
-	level.Info(logger).Log("msg", "created memcached index cache")
+	level.Info(logger).Log("msg", "created cacheClient index cache")
```
Comment (colega) [inline pkg/store/cache/memcached.go:56]: Nitpick: seems to be excessive renaming, maybe just `created index cache`?

File: pkg/cacheutil/redis_client.go (hunk 2 above)
Comment (hanjm) [reply pkg/cacheutil/redis_client.go:171]: let me fix it.

Comment (hanjm) [reply pkg/cacheutil/redis_client.go:171]: let me fix it.

File: pkg/store/cache/memcached.go (hunk 4 above)
Comment (hanjm) [reply pkg/store/cache/memcached.go:56]: Thank you very much. Done by #4931.

File: pkg/store/cache/caching_bucket_factory.go (hunk 3 above)
Comment (hanjm) [reply pkg/store/cache/caching_bucket_factory.go:104]: Thank you very much. Done by #4931.

**Comment by pgold30** [comment]
Love to see this, do you have a performance comparison with / without this to see if its better ?

**Comment by fagossa** [comment]
Hello,
any idea when this is going to be release? 
thanks a lot for the great job 👍

**Pull request merged**